pkg/types/dsse/v0.0.1/entry.go

```diff
@@ -0,0 +1,337 @@
+		Payload:     string(v.DsseObj.Payload),
+		PayloadType: *v.DsseObj.PayloadType,
+	}
+
+	allPubKeyBytes := make([][]byte, 0)
+	for _, sig := range v.DsseObj.Signatures {
+		env.Signatures = append(env.Signatures, dsse.Signature{
+			KeyID: sig.Keyid,
+			Sig:   string(sig.Sig),
+		})
+
+		allPubKeyBytes = append(allPubKeyBytes, sig.PublicKey)
+	}
+
+	if _, err := verifyEnvelope(allPubKeyBytes, env); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v *V001Entry) Canonicalize(ctx context.Context) ([]byte, error) {
+	canonicalEntry := models.DsseV001Schema{
+		PayloadHash: v.DsseObj.PayloadHash,
+		Signatures:  v.DsseObj.Signatures,
+		PayloadType: v.DsseObj.PayloadType,
+	}
+
+	model := models.Dsse{}
+	model.APIVersion = swag.String(APIVERSION)
+	model.Spec = canonicalEntry
+	return json.Marshal(&model)
+}
+
+func (v *V001Entry) Attestation() []byte {
+	payload := v.DsseObj.Payload
+	if len(payload) > viper.GetInt("max_attestation_size") {
+		log.Logger.Infof("Skipping attestation storage, size %d is greater than max %d", len(payload), viper.GetInt("max_attestation_size"))
+		return nil
+	}
+
+	return payload
+}
+
+type verifier struct {
+	v signature.Verifier
+}
+
+func (v *verifier) KeyID() (string, error) {
+	return "", nil
+}
+
+func (v *verifier) Public() crypto.PublicKey {
+	// the dsse library uses this to generate a key ID if the KeyID function returns an empty string
+	// as well for the AcceptedKey return value.  Unfortunately since key ids can be arbitrary, we don't
+	// know how to generate a matching id for the key id on the envelope's signature...
+	// dsse verify will skip verifiers whose key id doesn't match the signature's key id, unless it fails
+	// to generate one from the public key... so we trick it by returning nil ¯\_(ツ)_/¯
+	return nil
+}
+
+func (v *verifier) Verify(data, sig []byte) error {
+	if v.v == nil {
+		return errors.New("nil verifier")
+	}
+	return v.v.VerifySignature(bytes.NewReader(sig), bytes.NewReader(data))
+}
+
+func (v V001Entry) CreateFromArtifactProperties(_ context.Context, props types.ArtifactProperties) (models.ProposedEntry, error) {
+	returnVal := models.Dsse{}
+	re := V001Entry{}
+
+	var err error
+	artifactBytes := props.ArtifactBytes
+	if artifactBytes == nil {
+		if props.ArtifactPath == nil {
+			return nil, errors.New("path to artifact file must be specified")
+		}
+		if props.ArtifactPath.IsAbs() {
+			return nil, errors.New("dsse envelopes cannot be fetched over HTTP(S)")
+		}
+		artifactBytes, err = ioutil.ReadFile(filepath.Clean(props.ArtifactPath.Path))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	env := dsse.Envelope{}
+	if err := json.Unmarshal(artifactBytes, &env); err != nil {
+		return nil, fmt.Errorf("payload must be a valid dsse envelope: %w", err)
+	}
+
+	allPubKeyBytes := make([][]byte, 0)
+	if props.PublicKeyBytes != nil {
+		allPubKeyBytes = append(allPubKeyBytes, props.PublicKeyBytes)
+	}
+
+	allPubKeyBytes = append(allPubKeyBytes, props.PublicKeysBytes...)
```

Comment:
I might be misunderstanding this, but it looks like the props.PublicKeyBytes will get appended twice. Is that intentional?

Reply:
PublicKeyBytes vs PublicKeysBytes

Part of the discussion on the issue was how to handle passing multiple public keys for cases where we would want to validate multiple signatures on the envelope.  To handle this a new prop called PublicKeysBytes was added to handle that case.  It's not a very nice solution though, so definitely open to alternatives here.